feat: add SHA-256 and SHA-512 wrapper implementations

Farukest opened the pull request:
## Describe your changes

Implements Hasher, HasherExt, and ElementHasher traits for SHA2 hash functions.

- SHA-256 with 128-bit collision resistance
- SHA-512 with 256-bit collision resistance
- NIST test vectors for both algorithms
- Property-based tests with proptest

Closes #689

## Checklist before requesting a review
- [x] Repo forked and branch created from \`next\` according to naming convention.
- [x] Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- [x] Relevant issues are linked in the PR description.
- [x] Tests added for new functionality.
- [x] Documentation/comments updated according to changes.

adr1anh commented on miden-crypto/src/hash/sha2/mod.rs:
```diff
@@ -0,0 +1,448 @@
+}
+
+/// Cast the slice into contiguous bytes.
+fn prepare_merge<const N: usize, D>(args: &[D; N]) -> &[u8]
```
This feels like a more generally useful function that we might want to make available/reuse across different modules. Was this copied from another hash function wrapper?

Farukest replied:
Yes, this was copied from the `blake` and `keccak` modules which have identical implementations. I've added a note to the function doc comment acknowledging this duplication and suggesting consolidation into a shared utility in the future.

adr1anh commented on miden-crypto/src/hash/sha2/mod.rs:
```diff
@@ -0,0 +1,448 @@
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+#[cfg_attr(feature = "serde", serde(into = "String", try_from = "&str"))]
+pub struct Sha256Digest([u8; DIGEST256_BYTES]);
```
I wonder if it would be cleaner to have `struct Sha2Digest<const D: usize>([u8; D])` with type alias `Sha256Digest = Sha2<256>`.

adr1anh replied:
I think it might be worth having a more general `Digest256` and `Digest512` struct which can be reused among the different binary hash functions. But let's create an issue and do this in a follow up

Al-Kindi-0 commented on miden-crypto/src/hash/sha2/tests.rs:
```diff
@@ -0,0 +1,212 @@
+// ================================================================================================
+
+// NIST test vectors for SHA-256
+// https://csrc.nist.gov/projects/cryptographic-standards-and-guidelines/example-values
```
nit: could we get a more precise link to the test vectors

Al-Kindi-0 commented on miden-crypto/src/hash/sha2/tests.rs:
```diff
@@ -0,0 +1,212 @@
+];
+
+// NIST test vectors for SHA-512
+// https://csrc.nist.gov/projects/cryptographic-standards-and-guidelines/example-values
```
nit: same point regarding the link

Farukest replied:
All Done @Al-Kindi-0

Al-Kindi-0 replied:
Thank you!

Al-Kindi-0 reviewed (approved):
Look good, thank you!

Farukest commented:
Hi @huitseeker, I received email notifications about your review comments
  ( regarding `#[repr(transparent)]`, memory layout tests, and SHA-512 truncation docs ),
  but I can't see them on GitHub. I couldn't be sure if it was about github comment load issue or anything else..

  In the meantime, I've addressed what I understood from the emails but did not push yet:
  - Added `#[repr(transparent)]` to digest types
  - Added memory layout and `digests_as_bytes` correctness tests
  - Added PropTest for `merge_many`
  - Documented SHA-512 truncation behavior (not SHA-512/256)
  - Added test demonstrating the difference
  
  @bobbinth @adr1anh @Al-Kindi-0

huitseeker reviewed (dismissed):
Actually, so sorry for the back and forth @Farukest but we won't be exposing truncated SHA512/256t, aka the truncated SHA-512 with the non-standard IVs, because it has zero usage footprint. We'd like to expose SHA-512/256 instead, which requires using the proper IVs. It's exposed in the `sha2` crate as [`Sha512_256`](https://docs.rs/sha2/latest/sha2/type.Sha512_256.html). Could you help make this change?

huitseeker commented:
@Farukest That sounds great. The cherry on top is if you could use my comment to open an issue on Winterfell, and refer to it in a line of code comment on SHA-512 in this PR explaining why we're not implementing `Digest` ,

Farukest commented:
I opened on winterfell @huitseeker and have one question that confuses my mind now.

The Hasher trait requires type `Digest: Digest`, so removing impl `Digest` for `Sha512Digest` also requires removing impl Hasher for `Sha512`. Should I remove both trait implementations 
and keep only standalone methods, or keep the truncated `Digest` implementation until winterfell/issues/406 is resolved?

Farukest commented:
> I would remove the trait implementations and keep the standalone methods. The traits are there for hash functions that we use for STARK proof generation, and SHA512 is not one of these hash functions (at least not in the near future) - so, we don't have to implement these traits for it.

Done @bobbinth 
@huitseeker
